File: src/php/Store/InMemoryStore.php
<?php

declare(strict_types=1);

namespace Anonympins\Fingerprint\Store;

/**
 * In-memory implementation of the IStore interface.
 * Suitable for local development, tests, or single-instance deployments.
 */
class InMemoryStore implements IStore
{
    /**
     * @var array<string, array{value: mixed, expiresAt: int|null}>
     */
    private array $data = [];

    public function get(string $key)
    {
        if (!isset($this->data[$key])) {
            return null;
        }

        $item = $this->data[$key];
        if ($item['expiresAt'] !== null && $item['expiresAt'] < time()) {
            $this->delete($key); // Delete expired item
            return null;
        }

        return $item['value'];
    }

    public function set(string $key, $value, ?int $ttl = null): void
    {
        $expiresAt = $ttl !== null ? time() + $ttl : null;

        $this->data[$key] = ['value' => $value, 'expiresAt' => $expiresAt];
    }

    public function has(string $key): bool
    {
        if (!isset($this->data[$key])) {
            return false;
        }

        $item = $this->data[$key];
        if ($item['expiresAt'] !== null && $item['expiresAt'] < time()) {
            $this->delete($key);
            return false;
        }

        return true;
    }

    public function delete(string $key): void
    {
        unset($this->data[$key]);
    }

    /**
     * Clears all data from the store. Useful for testing.
     */
    public function clear(): void
    {
        $this->data = [];
    }
}

File: src/java/src/com/anonympins/fingerprint/InMemoryStore.java
package com.anonympins.fingerprint;

import com.anonympins.fingerprint.utils.ChallengeUtils;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class InMemoryStore implements IStore {
    private static class StoreItem {
        final Object value;
        final Long expiresAt;

        StoreItem(Object value, Long expiresAt) {
            this.value = value;
            this.expiresAt = expiresAt;
        }

        boolean isExpired() {
            return expiresAt != null && expiresAt < System.currentTimeMillis();
        }
    }

    private final Map<String, StoreItem> data = new ConcurrentHashMap<>();

    public InMemoryStore() {
        ChallengeUtils.setStore(this);
    }

    @Override
    public Object get(String key) {
        StoreItem item = data.get(key);
        if (item == null) {
            return null;
        }
        if (item.isExpired()) {
            delete(key);
            return null;
        }
        return item.value;
    }

    @Override
    public void set(String key, Object value, Integer ttl) {
        Long expiresAt = ttl != null ? System.currentTimeMillis() + (ttl * 1000L) : null;
        data.put(key, new StoreItem(value, expiresAt));
    }

    @Override
    public boolean has(String key) {
        StoreItem item = data.get(key);
        if (item == null) {
            return false;
        }
        if (item.isExpired()) {
            delete(key);
            return false;
        }
        return true;
    }

    @Override
    public void delete(String key) {
        data.remove(key);
    }

    public void clear() {
        data.clear();
    }
}